### PR TITLE
fix: retry GitHub token for Spotty Bunny Tab completion

### LINKED ISSUE #356: Spotty Bunny: retry GitHub token resolution; surface and log auth failures
## Summary

Spotty Bunny repo Tab completion can stay broken for the lifetime of the LaunchAgent process after a single failed `gh auth token` lookup — even after the user fixes auth with `gh auth login`. Failures are also silent in the UI (empty completion table, no status text) and under-logged (Tab lines with no auth/API warning).

## Background / repro

Observed 2026-08-25:

1. `gh` auth was working in a fresh shell (`gh auth status` OK; `gh auth token` + API HTTP 200).
2. Spotty Bunny (LaunchAgent) had been running since **2026-08-22**.
3. The `gh:github.com` keychain item was recreated on **2026-08-24** (re-login), but the agent was never restarted.
4. Tab complete for repo params (`prh repo`, `pr dom`, …) produced no candidates; logs showed `tab complete …` with no `completion failed` warning.
5. A fresh Python process using the same installed package resolved a token and completed repos correctly.
6. Restarting the agent (`launchctl kickstart -k gui/$(id -u)/com.thehcma.bunnify.spotty-bunny`) restored completion.

Root cause in `app/github_complete.py` `resolve_github_token`: a failed `gh auth token` is stored in `_token_cache` as `None` and never retried for the process lifetime.

```python
if use_cache and _token_cache is not _UNSET:
    return _token_cache if isinstance(_token_cache, str) else None
token = github_token_from_gh(runner=runner)
if use_cache:
    _token_cache = token  # caches None forever after one miss
```

Spotty Bunny also never calls `bootstrap_github_completion_cache` / `ensure_github_authenticated` on completer load (the REPL CLI does), so disk cache stays empty and there is no startup auth status line.

## Expected behavior

1. **Automatic retry:** If `_token_cache` is a negative miss (`None`), periodically re-run `gh auth token` (and/or clear the negative cache after a TTL, e.g. 30–60s) so a later `gh auth login` is picked up without restarting Spotty Bunny.
2. **Positive cache OK:** Successful tokens can still be cached (with optional TTL / refresh).
3. **Surface errors in the search box:** When Tab is pressed for a GitHub-backed param and no token / API auth fails, set the Spotty Bunny status line under the field (same place as other guidance), e.g.:
   - `GitHub auth needed — run: gh auth login`
   - `GitHub API error: …` (short, fail-soft)
   Empty completion alone is not enough feedback.
4. **Log failures:** Write the same class of failures to the Spotty Bunny application log (`BUNNIFY_DATA_DIR/spotty-bunny.log`, default `~/.local/share/bunnify/spotty-bunny.log`) at **WARNING** (or higher), including enough context to diagnose without secrets:
   - missing / failed `gh auth token` (and whether negative cache skipped a re-probe)
   - GitHub REST failures used for completion (HTTP status / error class, path — not the token)
   - Tab completion aborted because auth was unavailable
   Do not log token values. Prefer a single clear warning per failure mode rather than only `logger.debug` (today many auth/API misses are debug-only and invisible at default INFO).

## Suggested implementation notes

- In `resolve_github_token`: do not permanently cache `None`; use a short negative TTL or always re-probe when cache is empty/`None`.
- Optionally clear `_token_cache` on successful `gh auth login` paths already handled by `ensure_github_authenticated`.
- In Spotty Bunny (`_load_completer` / completion path): resolve token (non-interactive), bootstrap/warm completion cache like the CLI, and on Tab when `suggest_param_values` returns `[]` because auth is missing, call `_set_status(...)` with actionable text (reuse `gh_install_guidance` / CLI messaging where possible) **and** `logger.warning(...)`.
- Raise or keep `github_complete` auth/API miss logging at WARNING when it affects interactive completion (throttle if needed so retries do not spam).
- Add unit tests for negative-cache expiry, status messaging when token resolve fails, and that failure helpers emit a warning-level log record.

## Workaround

Restart Spotty Bunny after fixing auth:

```bash
launchctl kickstart -k "gui/$(id -u)/com.thehcma.bunnify.spotty-bunny"
# or: spotty-bunny upgrade
```

## Acceptance

- [ ] After a failed token lookup, a later successful `gh auth login` is picked up within ~1 minute (or on next Tab) without restarting the agent.
- [ ] When auth is still broken, Tab on a repo/PR/issue param shows a clear status message in the search UI.
- [ ] The same failure is visible in `spotty-bunny.log` at WARNING+ (no token material in the log line).
- [ ] Tests cover negative-cache retry, UI/status surfacing, and warning-level logging (or the pure helpers that drive it).

### PR DESCRIPTION
**Pull request by thehcma**
## Summary
- Expire negative `gh auth token` cache (~60s TTL) so Spotty Bunny picks up a later `gh auth login` without restart
- Surface missing GitHub auth in the search status line when Tab hits a GitHub-backed param
- Log auth/API completion failures at WARNING (throttled, no token material) and bootstrap the completion cache on panel show
- Closes #356

## Test plan
- [x] `uv run ruff check .` / `ruff format --check .` / `pyright --warnings`
- [x] `./test_bunnify` (includes new negative-TTL and blocked-message tests)
- [x] `./test_integration`
- [ ] After install/restart: with no auth, Tab on `prh ` shows status + WARNING in `spotty-bunny.log`
- [ ] `gh auth login`, wait ≤60s (or Tab again after TTL), completions work without agent restart

### CODE REVIEW

File: app/spotty_bunny_app.py [hunk 1]
```diff
@@ -984,7 +1001,19 @@ def _show_completions(self, rows: list[CompletionRow]) -> None:
         self._completion_rows = rows
         if not rows:
             self._hide_completions()
+            message = github_param_completion_blocked_message(
```
[the remark this answers is not in the document]
Comment (thehcma) [reply app/spotty_bunny_app.py]: Fixed in 87ecb5d: capture `prefix = self._completion_prefix` before `_hide_completions()`, then pass that local into `github_param_completion_blocked_message` / the WARNING log. Added a regression test that asserts this ordering in `_show_completions`.

**Comment by thehcma** [comment]
Fixed in 87ecb5d: same as the inline thread — preserve the Tab prefix before hide so the auth status line and WARNING actually fire.

(In reply to https://github.com/the-hcma/bunnify/pull/357#pullrequestreview-5019523146)

**Comment by thehcma** [comment]
Fixed in 87ecb5d: capture Tab prefix before _hide_completions so auth status/WARNING fire. (In reply to https://github.com/the-hcma/bunnify/pull/357#pullrequestreview-5019523146)

(In reply to https://github.com/the-hcma/bunnify/pull/357#pullrequestreview-5019523146)

File: app/spotty_bunny_app.py [hunk 2]
```diff
@@ -983,8 +1000,21 @@ def _schedule_update_check(self) -> None:
     def _show_completions(self, rows: list[CompletionRow]) -> None:
         self._completion_rows = rows
         if not rows:
+            prefix = self._completion_prefix
             self._hide_completions()
+            message = github_param_completion_blocked_message(
```
[the remark this answers is not in the document]
Comment (thehcma) [reply app/spotty_bunny_app.py]: Fixed in eef913b: Tab completion worker now returns `(rows, blocked_message)` so `gh auth token` / keychain runs on the IO thread; `_show_completions` only applies the precomputed status on the main thread.

File: bookmarks/test_spotty_bunny.py [hunk 3]
```diff
@@ -1425,6 +1425,105 @@ def test_apply_completion_replaces_prefix(self) -> None:
...
+            )
+
+        with self.assertLogs("app.github_complete", level=logging.WARNING):
+            message = github_param_completion_blocked_message(
+                "prh dom",
+                entries,
+                environ={"PATH": "/usr/bin"},
+                runner=gh_runner,
+            )
+        self.assertEqual(message, GITHUB_AUTH_NEEDED_MESSAGE)
```
[the remark this answers is not in the document]
Comment (thehcma) [reply bookmarks/test_spotty_bunny.py:1471]: Fixed in eef913b: patched `gh_is_available` to True in the unauthenticated-message test so the assertion is host-independent.

**Comment by thehcma** [comment]
Fixed in eef913b: (1) blocked-message + token resolve moved off the UI thread with completions; (2) unauth test mocks `gh_is_available`. (In reply to https://github.com/the-hcma/bunnify/pull/357#pullrequestreview-5019695995)

(In reply to https://github.com/the-hcma/bunnify/pull/357#pullrequestreview-5019695995)

File: app/spotty_bunny_app.py [hunk 4]
```diff
@@ -980,11 +1008,25 @@ def _schedule_update_check(self) -> None:
...
+        self,
+        rows: list[CompletionRow],
+        *,
+        blocked_message: str | None = None,
+    ) -> None:
         self._completion_rows = rows
         if not rows:
+            prefix = self._completion_prefix
             self._hide_completions()
+            if blocked_message is not None:
```
[the remark this answers is not in the document]
Comment (thehcma) [reply app/spotty_bunny_app.py]: Fixed in d197217: extracted `surface_blocked_github_completion` and added unit tests that assert status + throttled WARNING (including a second-call throttle no-op).

File: app/spotty_bunny_app.py [hunk 5]
```diff
@@ -980,11 +1008,25 @@ def _schedule_update_check(self) -> None:
...
+                logger.warning(
```
[the remark this answers is not in the document]
Comment (thehcma) [reply app/spotty_bunny_app.py]: Fixed in d197217: blocked Tab and startup auth messages now go through `warn_github_completion` (`_warn_throttled`) so repeated Tabs do not spam the log.

File: app/github_complete.py [hunk 6]
```diff
@@ -37,11 +38,16 @@
 _ISSUE_PARAM_NAMES = frozenset({"issue", "issue_id", "issue_num", "issue_number"})
 _API_ROOT = "https://api.github.com"
 _UNSET = object()
+_WARN_THROTTLE_SECONDS = 60.0
+
+GITHUB_AUTH_NEEDED_MESSAGE = "GitHub auth needed — run: gh auth login"
```
Comment (thehcma) [inline app/github_complete.py:43]: this error should only surface if the current set of bookmarks is using anything GitHub-related in its tab completion

Comment (thehcma) [reply app/github_complete.py:43]: Fixed in $(git rev-parse --short HEAD): auth WARNINGs / gh-install offer / cache warm now run only when loaded bookmarks declare GitHub-backed Tab completion (`entries_need_github_completion`). Tab-time status was already gated on the active GitHub param slot.

**Comment by thehcma** [comment]
bugbot run

**Comment by thehcma** [comment]
Acknowledged — Bugbot is disabled for this repo; continuing with the remaining review agents. (In reply to https://github.com/the-hcma/bunnify/pull/357#issuecomment-5412101623)

(In reply to https://github.com/the-hcma/bunnify/pull/357#issuecomment-5412101623)

File: app/github_complete.py [hunk 7]
```diff
@@ -257,6 +301,23 @@ def github_token_from_environ(
...
+    complete: Mapping[str, ParamCompleteSpec] | None = None,
+) -> bool:
+    """True when *param_name* is completed via the GitHub REST API."""
+    if complete is not None:
+        spec = complete.get(param_name)
+        if spec is not None:
+            return True
+    name = param_name.lower()
+    return (
+        name in _REPO_PARAM_NAMES
```
[the remark this answers is not in the document]
Comment (thehcma) [reply app/github_complete.py:314]: Fixed in $(git rev-parse --short HEAD): when a `complete` map is attached (including empty), only declared specs count for GitHub auth UX. Name heuristics no longer fire for bare `number`/`repo`/… params, so non-GitHub bookmarks stay quiet.

**Comment by thehcma** [comment]
Fixed in $(git rev-parse --short HEAD): GitHub auth UX gating now requires an explicit complete spec. (In reply to https://github.com/the-hcma/bunnify/pull/357#pullrequestreview-5020334444)

(In reply to https://github.com/the-hcma/bunnify/pull/357#pullrequestreview-5020334444)
